src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+use slice::from_raw_parts_mut;
+
+pub const DEFAULT_BUF_SIZE: usize = 64 * 1024;
+
+pub fn read_to_end_zeroed<R: Read + ?Sized>(r: &mut R, buf: &mut Vec<u8>) -> io::Result<usize> {
```

Comment:
To make these accept trait objects you may need to force an extra layer of indirection in callers. For example if you have `R: Read + ?Sized` then `&mut R` may already be a trait object, and we can't make a trait object out of a trait object.

You can get around this via:

```rust
fn foo<R: Read + ?Sized>(mut r: &mut R) {
    read_to_end(&mut r);
}
```